samples/matter/common/src/diagnostic/diagnostic_logs_end_user.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "diagnostic_logs_end_user.h"
+
+#include <lib/support/CodeUtils.h>
+#include <system/SystemError.h>
+#include <zephyr/retention/retention.h>
+
+using namespace chip;
+
+namespace
+{
+const struct device *kEndUserDataMem = DEVICE_DT_GET(DT_NODELABEL(end_user_logs_retention));
+} /* namespace */
+
+CHIP_ERROR DiagnosticLogsEndUser::Init()
+{
+	if (mIsInitialized) {
+		return CHIP_NO_ERROR;
+	}
+
+	mCapacity = retention_size(kEndUserDataMem);
+	VerifyOrReturnError(mCapacity > kHeaderSize, CHIP_ERROR_INTERNAL);
+
+	/* The actual capacity to store data is reduced by "header" tracking the stored size and data beginning. */
+	mCapacity = mCapacity - kHeaderSize;
+
+	int ret = retention_is_valid(kEndUserDataMem);
+
+	/* If data is invalid just clean the header data, as we are not sure which part is corrupted. */
+	if (ret != 1) {
+		ret = retention_write(kEndUserDataMem, 0, reinterpret_cast<uint8_t *>(&mCurrentSize),
+				      sizeof(mCurrentSize));
+		VerifyOrReturnError(ret == 0, System::MapErrorZephyr(ret));
+		ret = retention_write(kEndUserDataMem, sizeof(mCurrentSize), reinterpret_cast<uint8_t *>(&mDataBegin),
+				      sizeof(mDataBegin));
+		VerifyOrReturnError(ret == 0, System::MapErrorZephyr(ret));
+	} else {
+		/* Load the header data from retention RAM. */
+		ret = retention_read(kEndUserDataMem, 0, reinterpret_cast<uint8_t *>(&mCurrentSize),
+				     sizeof(mCurrentSize));
+		VerifyOrReturnError(0 == ret, System::MapErrorZephyr(ret));
+
+		ret = retention_read(kEndUserDataMem, sizeof(mCurrentSize), reinterpret_cast<uint8_t *>(&mDataBegin),
+				     sizeof(mDataBegin));
+		VerifyOrReturnError(0 == ret, System::MapErrorZephyr(ret));
+	}
+
+	mDataEnd = mDataBegin + mCurrentSize > mCapacity ? mCurrentSize + mDataBegin - mCapacity :
+							   mDataBegin + mCurrentSize;
+
+	mIsInitialized = true;
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DiagnosticLogsEndUser::Clear()
+{
+	return System::MapErrorZephyr(retention_clear(kEndUserDataMem));
+}
+
+CHIP_ERROR DiagnosticLogsEndUser::PushLog(const void *data, size_t size)
+{
+	VerifyOrReturnError(data, CHIP_ERROR_INVALID_ARGUMENT);
+	VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+	VerifyOrReturnError(size < mCapacity, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+	int ret = 0;
+
+	const size_t freeSpace = mCapacity - mCurrentSize;
+	/* There is no place for the new data. Forget the oldest data by moving the offset forward. */
```

Comment:
Can't we use the `ring_buffer` instead?

Reply:
We can't because the ring buffer operates on RAM, and does not support read/write operations to retention area.